drivers/hmis/app/models/hmis/form/definition.rb

```diff
@@ -57,6 +59,132 @@ def self.validate_json(json)
+    end
+
+    return [nil, errors.errors] if errors.errors.any?
+
+    # Additional validations for HUD assessment dates to be within appropriate entry/exit bounds
+    if date.present? && hud_assessment?
+      # Ensure assessment date is on or after entry date
+      errors.add item.field_name, :out_of_range, readable_attribute: readable_attribute, message: "must be after entry date (#{entry_date.strftime('%m/%d/%Y')})" if entry_date.present? && !intake? && date < entry_date
+      # Ensure assessment date is on or before exit date
+      errors.add item.field_name, :out_of_range, readable_attribute: readable_attribute, message: "must be before exit date (#{exit_date.strftime('%m/%d/%Y')})" if exit_date.present? && !exit? && date > exit_date
+    end
+
+    date = nil if errors.errors.any?
+
+    [date, errors.errors]
+  end
+
+  def validate_form_values(hud_values, _custom_values = nil)
+    errors = HmisErrors::CustomValidationErrors.new
+    hud_values.each do |link_id, value|
+      item = link_id_item_hash[link_id.to_s]
+      raise "Unrecognized link ID: #{link_id}" unless item.present?
+
+      next if item.assessment_date # Skip assessment date, it is validated separately
+
+      # Use item text as the "readable attribute" name to display on errors
+      readable_attribute = item.brief_text || item.text
+
+      is_missing = value.nil?
+      is_dnc = value == 'DATA_NOT_COLLECTED'
+
+      # Validate required status
+      # TODO: confirm logic
+      if item.required && is_missing
+        errors.add item.field_name, :required, readable_attribute: readable_attribute
+      elsif (item.warn_if_empty || item.field_name || item.required) && is_dnc
+        errors.add item.field_name, :data_not_collected, severity: :warning, readable_attribute: readable_attribute
+      end
+
+      # TODO: Evaluate EnableWhen and null out any disabled fields
+      # TODO: Validate ValuBounds
+      # TODO: Add support for RequiredWhen and implement it
+      # How to handle bounds that rely on local values like projectStartDate and entryDate?
+    end
+
+    errors.errors
+  end
+
+  def key_by_field_name(hud_values)
+    result = {}
+    recur_fill = lambda do |items, current_record_type|
+      items.each do |item|
+        if item.item
+          record_type = Types::Forms::Enums::RelatedRecordType.values[item.record_type]&.description if item.record_type.present?
+          recur_fill.call(item.item, record_type || current_record_type)
+        end
+
+        next unless item.field_name.present?
+        next unless hud_values.key?(item.link_id)
+
+        key = item.field_name
+        key = "#{current_record_type}.#{key}" if current_record_type.present?
+
+        result[key] = hud_values[item.link_id]
+      end
+    end
+
+    recur_fill.call(definition_struct.item, nil)
+    result
+  end
+
+  private def definition_struct
+    @definition_struct ||= JSON.parse definition, object_class: OpenStruct
```

Comment:
Sorry I didn't catch this earlier, we should probably be using `OJ.load` and `OJ.dump` since, at least historically, they've been way faster.
https://github.com/ohler55/oj